### DISCUSSION TITLE
Express middleware compatibility

### DISCUSSION THREAD

**Discussion by robokozo** [nitrojs/nitro#1858] [Q&A] [answered]
Does anyone have an example of wrapping an express middleware for compatibility with nitro/h3? My company exclusively relies on old express middleware that prevent me from writing new code with nitro. 

I believe defineEventHandlers `event` param has req and res available on it... but I'm not sure if those are compatible with express. For reference, express middlewares provide: `(req, res, next) => { next() }`

Would it be something like:

```
export default defineEventHandler(async (event) => {
  await new Promise((resolve) => {
    myExpressMiddleware(event.node.req, event.node.res, resolve);
  });
});
```

**Comment by robokozo** [comment]
It looks a little more involved. From what I see, in express the res object has a bunch of functions on it that would need to be adapted for nitro. 

Like a `res.send("foo")` in express would just be a `return "foo"` in nitro/h3

**Comment by peterroe** [comment, marked as answer]
Hi, `H3` is compatible with express middleware. For your case, you can do like this:
```ts
export default fromNodeMiddleware((req, res, next) => {
    myExpressMiddleware(req, res, next);
})
```
Or we could even make it simpler.

```ts
export default fromNodeMiddleware(myExpressMiddleware)
```

**Comment by robokozo** [comment]
@peterroe wow! even better! thank you <3
